src/operations/client_bulk_write/results_merger.ts

```diff
@@ -0,0 +1,96 @@
+import { type Document } from '../../bson';
+import { type ClientBulkWriteCursorResponse } from '../../cmap/wire_protocol/responses';
+import {
+  type ClientBulkWriteOptions,
+  type ClientBulkWriteResult,
+  type ClientDeleteResult,
+  type ClientInsertOneResult,
+  type ClientUpdateResult
+} from './common';
+
+/**
+ * Merges client bulk write cursor responses together into a single result.
+ * @internal
+ */
+export class ClientBulkWriteResultsMerger {
+  result: ClientBulkWriteResult;
+  options: ClientBulkWriteOptions;
+
+  /**
+   * Instantiate the merger.
+   * @param options - The options.
+   */
+  constructor(options: ClientBulkWriteOptions) {
+    this.options = options;
+    this.result = {
+      insertedCount: 0,
+      upsertedCount: 0,
+      matchedCount: 0,
+      modifiedCount: 0,
+      deletedCount: 0,
+      insertResults: undefined,
+      updateResults: undefined,
+      deleteResults: undefined
+    };
+
+    if (options.verboseResults) {
+      this.result.insertResults = new Map<number, ClientInsertOneResult>();
+      this.result.updateResults = new Map<number, ClientUpdateResult>();
+      this.result.deleteResults = new Map<number, ClientDeleteResult>();
+    }
+  }
+
+  /**
+   * Merge the results in the cursor to the existing result.
+   * @param response - The cursor response.
+   * @param documents - The documents in the cursor.
+   * @returns The current result.
+   */
+  merge(
+    operations: Document[],
+    response: ClientBulkWriteCursorResponse,
+    documents: Document[]
+  ): ClientBulkWriteResult {
+    // Update the counts from the cursor response.
+    this.result.insertedCount += response.insertedCount;
+    this.result.upsertedCount += response.upsertedCount;
+    this.result.matchedCount += response.matchedCount;
+    this.result.modifiedCount += response.modifiedCount;
+    this.result.deletedCount += response.deletedCount;
+
+    if (this.options.verboseResults) {
+      // Iterate all the documents in the cursor and update the result.
+      for (const document of documents) {
+        // Only add to maps if ok: 1
+        if (document.ok === 1) {
+          // Get the corresponding operation from the command.
+          const operation = operations[document.idx];
+          // Handle insert results.
+          if ('insert' in operation) {
+            this.result.insertResults?.set(document.idx, { insertedId: operation.document._id });
+          }
+          // Handle update results.
+          if ('update' in operation) {
+            const result: ClientUpdateResult = {
+              matchedCount: document.n,
+              modifiedCount: document.nModified ?? 0,
+              // We do specifically want to check undefined here since a null _id is valid.
+              // eslint-disable-next-line no-restricted-syntax
+              didUpsert: document.upserted?._id !== undefined
```

Comment:
I'm sorry I didn't ask before. Isn't the existence of `upserted` the condition we should use here?

```js
didUpsert: document.upserted != null
```

While this is also going to give the correct result it relies on our BSON library's implementation rather than the API the server promises which is to omit this field if no upsert was performed. In practice, `document.upserted` is going to missing so the `?.` will always short circuit before trying `._id`

Reply:
Yes that's correct. I've updated now.